File: java/java/Dart/JavaParserBase.dart
import 'package:antlr4/antlr4.dart';
import 'JavaParser.dart';

abstract class JavaParserBase extends Parser {
    JavaParserBase(TokenStream input) : super(input);

    bool DoLastRecordComponent() {
        final ctx = this.context;
        if (ctx is! RecordComponentListContext) return true;
        final rcs = ctx.children
                ?.whereType<RecordComponentContext>()
                .toList() ?? [];
        if (rcs.isEmpty) return true;
        final count = rcs.length;
        for (int c = 0; c < count; ++c) {
            if (rcs[c].ELLIPSIS() != null && c + 1 < count) return false;
        }
        return true;
    }

    bool IsNotIdentifierAssign() {
        final identifierLikeTokens = {
            JavaParser.TOKEN_IDENTIFIER,
            JavaParser.TOKEN_MODULE,
            JavaParser.TOKEN_OPEN,
            JavaParser.TOKEN_REQUIRES,
            JavaParser.TOKEN_EXPORTS,
            JavaParser.TOKEN_OPENS,
            JavaParser.TOKEN_TO,
            JavaParser.TOKEN_USES,
            JavaParser.TOKEN_PROVIDES,
            JavaParser.TOKEN_WHEN,
            JavaParser.TOKEN_WITH,
            JavaParser.TOKEN_TRANSITIVE,
            JavaParser.TOKEN_YIELD,
            JavaParser.TOKEN_SEALED,
            JavaParser.TOKEN_PERMITS,
            JavaParser.TOKEN_RECORD,
            JavaParser.TOKEN_VAR,
        };
        final la = tokenStream.LA(1);
        if (!identifierLikeTokens.contains(la)) return true;
        final la2 = tokenStream.LA(2);
        return la2 != JavaParser.TOKEN_ASSIGN;
    }
}

File: java/java/TypeScript/JavaParserBase.ts
import { Parser, TokenStream, BufferedTokenStream } from 'antlr4';
import JavaParser, { RecordComponentContext, RecordComponentListContext } from './JavaParser.js';

export default abstract class JavaParserBase extends Parser {

    constructor(input: TokenStream) {
        super(input);
    }

    public DoLastRecordComponent(): boolean {
        const ctx = this._ctx;
        if (!(ctx instanceof RecordComponentListContext)) return true;
        const rcs = (ctx.children ?? []).filter(c => c instanceof RecordComponentContext) as RecordComponentContext[];
        if (rcs.length === 0) return true;
        const count = rcs.length;
        for (let c = 0; c < count; ++c) {
            if (rcs[c].ELLIPSIS() !== null && c + 1 < count) return false;
        }
        return true;
    }

    public IsNotIdentifierAssign(): boolean {
        const stream = this._input as BufferedTokenStream;
        const la = stream.LA(1);
        switch (la) {
            case JavaParser.IDENTIFIER:
            case JavaParser.MODULE:
            case JavaParser.OPEN:
            case JavaParser.REQUIRES:
            case JavaParser.EXPORTS:
            case JavaParser.OPENS:
            case JavaParser.TO:
            case JavaParser.USES:
            case JavaParser.PROVIDES:
            case JavaParser.WHEN:
            case JavaParser.WITH:
            case JavaParser.TRANSITIVE:
            case JavaParser.YIELD:
            case JavaParser.SEALED:
            case JavaParser.PERMITS:
            case JavaParser.RECORD:
            case JavaParser.VAR:
                break;
            default:
                return true;
        }
        const la2 = stream.LA(2);
        if (la2 !== JavaParser.ASSIGN) return true;
        return false;
    }
}
